python/tests/test_types.py

```diff
@@ -204,3 +206,388 @@ def test_non_parameterized_type_equality(input_index, input_type, check_index, c
+    ).json()
+    expected = (
+        '{"type": "struct", "fields": ['
+        '{"id": 1, "name": "required_field", "type": "string", "required": true, "doc": "this is a doc"}, '
+        '{"id": 2, "name": "optional_field", "type": "int", "required": true}'
+        "]}"
+    )
+    assert actual == expected
+
+
+def test_deserialization_struct():
+    actual = StructType.parse_raw(
+        """
+    {
+        "type": "struct",
+        "fields": [{
+                "id": 1,
+                "name": "required_field",
+                "type": "string",
+                "required": true,
+                "doc": "this is a doc"
+            },
+            {
+                "id": 2,
+                "name": "optional_field",
+                "type": "int",
+                "required": true,
+                "doc": null
+            }
+        ]
+    }
+    """
+    )
+
+    expected = StructType(
+        NestedField(1, "required_field", StringType(), True, "this is a doc"), NestedField(2, "optional_field", IntegerType())
+    )
+
+    assert actual == expected
+
+
+def test_str_struct(simple_struct: StructType):
+    assert str(simple_struct) == "struct<1: required_field: required string (this is a doc), 2: optional_field: required int>"
+
+
+def test_repr_struct(simple_struct: StructType):
+    assert (
+        repr(simple_struct)
+        == "StructType(fields=[NestedField(field_id=1, name='required_field', field_type=StringType(), required=True), NestedField(field_id=2, name='optional_field', field_type=IntegerType(), required=True)])"
+    )
+
+
+def test_serialization_list(simple_list: ListType):
+    actual = simple_list.json()
+    expected = '{"type": "list", "element-id": 22, "element": "string", "element-required": true}'
+    assert actual == expected
+
+
+def test_deserialization_list(simple_list: ListType):
+    actual = ListType.parse_raw('{"type": "list", "element-id": 22, "element": "string", "element-required": true}')
+    assert actual == simple_list
+
+
+def test_str_list(simple_list: ListType):
+    assert str(simple_list) == "list<string>"
+
+
+def test_repr_list(simple_list: ListType):
+    assert repr(simple_list) == "ListType(type='list', element_id=22, element_type=StringType(), element_required=True)"
+
+
+def test_serialization_map(simple_map: MapType):
+    actual = simple_map.json()
+    expected = """{"type": "map", "key-id": 19, "key": "string", "value-id": 25, "value": "double", "value-required": false}"""
+
+    assert actual == expected
+
+
+def test_deserialization_map(simple_map: MapType):
+    actual = MapType.parse_raw(
+        """{"type": "map", "key-id": 19, "key": "string", "value-id": 25, "value": "double", "value-required": false}"""
+    )
+    assert actual == simple_map
+
+
+def test_str_map(simple_map: MapType):
+    assert str(simple_map) == "map<string, double>"
+
+
+def test_repr_map(simple_map: MapType):
```

Comment:
It would be nice to have a more complex type that mixes maps, lists, and structs within one another, but I think this is good overall.